File: backend/app/models/dietary_preference.py
from __future__ import annotations

from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, Integer, JSON, UniqueConstraint, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database.base import Base


class DietaryPreference(Base):
    __tablename__ = "dietary_preferences"
    __table_args__ = (UniqueConstraint("user_id", name="uq_dietary_preferences_user"),)

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    user_id: Mapped[int] = mapped_column(
        ForeignKey("users.id"),
        index=True,
        nullable=False,
    )
    liked: Mapped[list[str]] = mapped_column(JSON, nullable=False, default=list)
    disliked: Mapped[list[str]] = mapped_column(JSON, nullable=False, default=list)
    avoid: Mapped[list[str]] = mapped_column(JSON, nullable=False, default=list)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    user = relationship("User", back_populates="dietary_preference")

File: frontend/types/dietary-preference.ts
export type DietaryPreference = {
  id: number;
  user_id: number;
  liked: string[];
  disliked: string[];
  avoid: string[];
  created_at: string;
  updated_at: string;
};

export type DietaryPreferencePayload = Pick<
  DietaryPreference,
  "liked" | "disliked" | "avoid"
>;

export type DietaryPreferenceResponse = {
  preference: DietaryPreference | null;
};
